### PR TITLE
[rush] Install prettier in postRushInstall hook

### PR DESCRIPTION
**Pull request by Claudiazhaoya**
## Summary

This PR aims to remove installing prettier from pre-commit hook and change it to be installed after rush install is done. 
## Details

Use 'npm install prettier' in postRushInstall hook and only run prettier in git hooks.

### CODE REVIEW

File: rush.json [hunk 1]
```diff
@@ -314,7 +314,7 @@
     /**
      * The list of shell commands to run after the Rush installation finishes
      */
-    "postRushInstall": [],
+    "postRushInstall": ["npm i -g --save-dev prettier || exit $?"],
```
Comment (dmichon-msft) [inline rush.json:317]: This seems like a slightly odd combination of flags.

Comment (Claudiazhaoya) [reply rush.json:317]: would you think 'npm install -g prettier' is better?

File: common/git-hooks/pre-commit [hunk 2]
```diff
@@ -6,4 +6,4 @@
 # Invoke the "rush prettier" custom command to reformat files whenever they
 # are committed. The command is defined in common/config/rush/command-line.json
 # and uses the "rush-prettier" autoinstaller.
-node common/scripts/install-run-rush.js prettier || exit $?
+prettier --write --ignore-unknown . || exit $?
```
Comment (dmichon-msft) [inline common/git-hooks/pre-commit:9]: This runs prettier on every file in the repository (less those excluded by prettierignore), which might work in a small repo (of which rushstack is reasonably so), but won't really scale well.

File: rush.json (hunk 1 above)
Comment (dmichon-msft) [reply rush.json:317]: The `--save-dev` and `--global` shouldn't be combined, so yes, that'd be better.

File: common/git-hooks/pre-commit (hunk 2 above)
Comment (Claudiazhaoya) [reply common/git-hooks/pre-commit:9]: That's what I noticed too. I think what we want is to run prettier on changed files only. So I guess we need to retrieve the changed files from staged files?

Comment (dmichon-msft) [reply common/git-hooks/pre-commit:9]: Unfortunately the prettier CLI doesn't support passing paths from stdin, and if you specify exact paths on the command line, the prettierignore seems to get bypassed. Ultimately this means we have to wrap it in a node script that calls the API, at which point it is probably easier to just have a package in the repository that gets installed as part of normal `rush install` and has the dependency on `prettier`. Then the hook can invoke the node script directly and error if `prettier` has not yet been installed.

File: rush.json (hunk 1 above)
Comment (iclanton) [reply rush.json:317]: A repo shouldn't be installing something globally. We also need to ensure that a consistent version is installed.

File: common/changes/@rushstack/heft/zhas-add-post-hook-to-install-prettier_2021-04-21-22-07.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (iclanton) [inline common/changes/@rushstack/heft/zhas-add-post-hook-to-install-prettier_2021-04-21-22-07.json:1]: Delete this file.

File: common/changes/@rushstack/node-core-library/zhas-add-post-hook-to-install-prettier_2021-04-21-22-07.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (iclanton) [inline common/changes/@rushstack/node-core-library/zhas-add-post-hook-to-install-prettier_2021-04-21-22-07.json:1]: Delete this file.

**Review by iclanton — CHANGES_REQUESTED** [review]

**Comment by iclanton** [comment]
I don't think this is the right approach. It would be better to run `npm install` in the autoinstaller folder and change the git hook to directly call `pretty-quick` in that folder, assuming that it's installed.

A better long-term approach would be to drop a config file (maybe `common/git-hooks/git-hooks-meta.json`) that tells Rush that the pre-commit hook depends on a particular autoinstaller, so Rush should install the autoinstaller when the hook is installed. Rush could just run `pnpm install` in the autoinstaller folder while running a normal install. It's not a large installation footprint, so it shouldn't have any noticeable performance impact on the install.

**Comment by octogonz** [comment]
> This PR aims to remove installing prettier from `pre-commit` hook and change it to be installed after rush install is done.

What happens if I run `git commit` without running `rush install`?

**Comment by iclanton** [comment]
Let's close this for now.

**Pull request closed without merging**